incubator/binding-mqtt/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -1983,28 +1983,24 @@ private void onDecodeSubscribe(
                 if (session)
                 {
                     final MqttSessionStateFW.Builder sessionStateBuilder =
-                        mqttSessionStateFW.wrap(sessionStateBuffer, 0, sessionStateBuffer.capacity())
-                            .clientId(clientId);
+                        mqttSessionStateFW.wrap(sessionStateBuffer, 0, sessionStateBuffer.capacity());
 
-                    newSubscriptions.forEach(subscription ->
+                    List<Subscription> currentSubscriptions = new ArrayList<>(sessionStream.subscriptions);
+                    currentSubscriptions.addAll(newSubscriptions);
+                    sessionStream.unAckedSubscriptions.addAll(newSubscriptions);
+                    currentSubscriptions.forEach(subscription ->
+                        sessionStateBuilder.subscriptionsItem(subscriptionBuilder ->
                         {
-                            sessionStateBuilder.subscriptionsItem(subscriptionBuilder ->
-                            {
-                                subscriptionBuilder.subscriptionId(subscription.id);
-                                subscriptionBuilder.flags(subscription.flags);
-                                subscriptionBuilder.pattern(subscription.filter);
-                            });
-                            sessionStream.unAckedSubscriptions.add(subscription);
-                        }
+                            subscriptionBuilder.subscriptionId(subscription.id);
+                            subscriptionBuilder.flags(subscription.flags);
+                            subscriptionBuilder.pattern(subscription.filter);
+                        })
                     );
 
                     final MqttSessionStateFW sessionState = sessionStateBuilder.build();
                     final int payloadSize = sessionState.sizeof();
 
-                    //TODO: is this correct? What is this?
-                    int reserved = payloadSize;
-
-                    sessionStream.doSessionData(traceId, reserved, sessionState);
+                    sessionStream.doSessionData(traceId, payloadSize, sessionState);
```

Comment:
`reserved` is `payload.sizeof() + padding`, so we do need to calculate `reserved` instead of assuming that `padding` is `0`.

Reply:
Let's discuss this as well. We probably need replyPad as well for decoding network data, but I'd like to understand padding more, before doing this.